docs/features/source-generators.md

```diff
@@ -0,0 +1,285 @@
+files we need to know which is which so we can know which file the breakpoints go with. A source generator of course is 
+allowed to stop emitting a file if needs change (if you delete a `.resx`, then the generated file associated with it 
+will also go away), but this gives us some control here.
+
+This was called "hint" in that the compiler is implicitly allowed to control the filename in however it ultimately 
+needs, and if two source generators give the same "hint" it can still distinguish them with any sort of 
+prefix/suffix as necessary.
+
+### IDE Integration
+
+One of the more complicated aspects of supporting generators is enabling a high-fidelity 
+experience in Visual Studio. For the purposes of determining code correctness, it is 
+expected that all generators will have had to be run. Obviously, it is impractical to run 
+every generator on every keystroke, and still maintain an acceptable level of performance 
+within the IDE.
+
+#### Progressive complexity opt-in
+
+It is expected instead that source generators would work on an 'opt-in' approach to IDE 
+enablement. 
+
+By default, a generator implementing only `ISourceGenerator` would see no IDE integration
+and only be correct at build time. Based on conversations with 1st party customers, 
+there are several cases where this would be enough.
+
+However, for scenarios such as code first gRPC, and in particular Razor and Blazor, 
+the IDE will need to be able to generate code on-they-fly as those file types are 
+edited and reflect the changes back to other files in the IDE in near real-time. 
+
+The proposal is to have a set of advanced interfaces that can be optionally implemented,
+that would allow the IDE to query the generator to decide what needs to be run in the case
+of any particular edit.
+
+For example an extension that would cause generation to run after saving a third party
+file might look something like: 
+
+```csharp
+namespace Microsoft.CodeAnalysis
+{
+    public interface ITriggeredByAdditionalFileSavedGenerator : ISourceGenerator
+    {
+        ImmutableArray<string> SupportedAdditionalFileExtensions { get; }
+    }
+}
+```
+It is expected that there will be various levels of opt in, that can be added to a generator
+in order to achieve the specific level of performance required of it. 
+
+What these exact APIs will look like remains an open question, and it's expected that we will 
+need to prototype some real-world generators before knowing what their precise shape will be.
+
+### Output files
+
+It it desirable that the generated source texts be available for inspection after generation,
+either as part of creating a generator or seeing what code was generated by a third party
+generator.
+
+By default, generated texts will be persisted to a `GeneratedFiles/{GeneratorAssemblyName}` 
+sub-folder within `CommandLineArguments.OutputDirectory`. The `fileNameHint` from 
```

Comment:
❔ Is `OutputDirectory` going to be under **obj** or **bin** for default scenarios?

💡 It would help implementers relate to this description to include an example of the output location.